src/Microsoft.AspNet.Server.Kestrel/Http/MemoryPool2.cs

```diff
@@ -0,0 +1,107 @@
+﻿using System;
+using System.Collections.Concurrent;
+
+namespace Microsoft.AspNet.Server.Kestrel.Http
+{
+    public class MemoryPool2 : IDisposable
```

Comment:
A brief description of what the algorithm used would be awesome. Bonus points if you add a why